### DISCUSSION TITLE
Tokio 1.4 Async TCPStream Connection `No Reactor` Error

### DISCUSSION THREAD

**Discussion by Evilsmaher** [tokio-rs/tokio#3631] [Q&A] [answered]
I have the following :

```
use tokio::net::TcpStream;

#[derive(Debug, Clone)]
enum ConnectionError {
	ConnectError
}

#[derive(Debug, Clone)]
struct Connection {
	stream: Arc<Option<TcpStream>>,
}

impl Connection {
	async fn connect() -> Result<Connection, ConnectionError> {
		println!("Connecting");
		
		let mut stream = TcpStream::connect("3.85.107.207:80").await.map_err(|_| ConnectionError::ConnectError)?;

		Ok( Connection {
			stream: Arc::new(Some(stream)),
			}
		)
	}
}
```

but when I run this function, it gives me the error

>thread '<unnamed>' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', C:\Users\bratl\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\context.rs:18:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

I'm not really sure how to adjust to this. I've considered 

```
Tokio::Spawn( )
```

but how do I get the return `stream` from that to then create a `Connection`?

**Comment by Darksonn** [comment]
Are you not using Tokio as your executor?

**Reply by Evilsmaher** [reply]
Dur, I see. I am using [Iced](https://github.com/hecrj/iced/tree/c1f70f1e9252b1971c17cd385273460c669fac26) and was using their default Executor... 

Now to figure out how to use the Tokio Executor with it :P

**Reply by Evilsmaher** [reply]
@Darksonn - I have included the Tokio Executor, it sees it, but am still getting a panic.

>thread 'tokio-runtime-worker' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', C:\Users\bratl\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\context.rs:18:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Cargo.toml**
```
[package]
name = "ClientAdminRust"
version = "0.1.0"
authors = ["bratl"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
iced = { version = "0.2", features = ["tokio"] }
rand = "0.8.0"
iced_native = "0.3"
tokio = { version = "1", features = ["full"] }
mini-redis = "0.4"
```

**Comment by taiki-e** [comment, marked as answer]
iced 0.2 uses [tokio 0.2 (and maybe tokio 0.3)](https://github.com/hecrj/iced/blob/futures-0.2.0/futures/Cargo.toml#L22). You have to use the iced's master branch that uses tokio 1.x, or tokio 0.2/0.3 that compatible with iced 0.2.

**Reply by Evilsmaher** [reply]
I tried running with tokio 0.2 and get the same issue.

```
[package]
name = "ClientAdminRust"
version = "0.1.0"
authors = ["bratl"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
iced = { version = "0.2", features = ["tokio"] }
rand = "0.8.0"
tokio = { version = "0.2", features = ["full"] }
mini-redis = "0.4"
```

>thread 'tokio-runtime-worker' panicked at 'there is no reactor running, must be called from the context of a Tokio 0.2.x runtime', C:\Users\bratl\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-0.2.25\src\io\driver\mod.rs:203:30
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Reply by taiki-e** [reply]
Could you try tokio_old feature of iced?

**Reply by Evilsmaher** [reply]
Wow - you are a beautiful, beautiful human being. I saw the `Executor`s in the iced files - for some reason didn't make the click to choose `tokio_old` when converting to 0.2. Thank you very much!

For anyone else:

```
[package]
name = "ClientAdminRust"
version = "0.1.0"
authors = ["bratl"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
iced = { version = "0.2", features = ["tokio_old"] }
rand = "0.8.0"
tokio = { version = "0.2", features = ["full"] }
mini-redis = "0.4"
```
